tools/generate_config.py

```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python3
+
+import argparse
+from pathlib import Path
+import json
+try:
+    import jsonschema
+except ImportError:
+    jsonschema = None
+from collections import OrderedDict
+
+THIS_DIR = Path(__file__).parent
+
+if (THIS_DIR / 'schemas').exists():
+    SCHEMAS_DIR = THIS_DIR / 'schemas'
+else:
+    SCHEMAS_DIR = THIS_DIR.parent / 'share' / 'schemas'
+
+ENCLAVE_CFG_SCHEMA_PATH = SCHEMAS_DIR / 'enclave-config.schema.json'
+
+# Validates a particular enclave_cfg_path against ENCLAVE_CFG_SCHEMA_PATH
+def validate(args):
+    if jsonschema is None:
+        raise RuntimeError('validate requires the "jsonschema" Python package')
+    schema_path = ENCLAVE_CFG_SCHEMA_PATH
+    print('Validating {args.file}')
+    with open(args.file) as f:
+      cfg = json.load(f)
+    with open(schema_path) as f:
+      cfg_schema = json.load(f)
+    jsonschema.validate(instance=cfg, schema=cfg_schema)
+    print('No errors found.')
+
+def post_type(jtype):
+  if 'type' in jtype:
+    jtt = jtype['type']
+    if 'maxLength' in jtype:
+      if jtt == 'string':
+        return '[' + str(jtype['maxLength']+1) + ']'
+      elif jtt == 'array':
+        return '[' + str(jtype['maxLength']) + ']'
+    return ''
+  else:
+    return ''
+
+def pre_type(jtype):
+  if 'type' in jtype:
+    jtt = jtype['type']
+    if jtt == 'array' or (isinstance(jtt, list) and 'array' in jtt and 'null' in jtt):
+      items = jtype['items']
+      item_type = pre_type(items)
+      if 'maxLength' in jtype:
+        return item_type
+      else:
+        return item_type + '*'
+    elif jtt == 'boolean': return 'bool'
+    elif jtt == 'none': return 'null'
+    elif jtt == 'string' or (isinstance(jtt, list) and 'string' in jtt and 'null' in jtt):
+      return 'char' if 'maxLength' in jtype else 'char*'
+    else:
+        print('unhandled json type: %s' % jtype)
```

Comment:
Fail hard? Also, try to use string interpolation like `f'unhandled json type: {jtype}'`.

Reply:
That's called 'interpolation'?! 

I used the same (and usual) style everywhere, it doesn't make sense to change that now. We can think about it if we ever extend this script.

Reply:
The `%` style is Python 2. Don't you think f-strings make it easier to read?

Reply:
No, to me they are not easier to read at all, but I grew up with `printf`. Also, a variable name appearing in a flat string without supposed semantics? Typos being detected at runtime instead of compile time? A flurry of string-prefixes like `f""` or `r""`, which are designed not to be noticed so they are better at hiding bugs? No, it's not a secret that I don't like Python :-)